R/stat-density-2d.r

```diff
@@ -10,34 +17,51 @@
 #'    using the a bandwidth estimator. For example, `adjust = 1/2` means
 #'    use half of the default bandwidth.
 #' @section Computed variables:
-#' Same as [stat_contour()]
-#'
-#' With the addition of:
+#' `stat_density_2d()` computes different variables depending on whether
+#' contouring is turned on or off. With contouring off (`contour = FALSE`),
+#' the following variables are provided:
 #' \describe{
-#'   \item{density}{the density estimate}
-#'   \item{ndensity}{density estimate, scaled to maximum of 1}
+#'   \item{`density`}{The density estimate.}
+#'   \item{`ndensity`}{Density estimate, scaled to a maximum of 1.}
+#'   \item{`count`}{Density estimate * number of observations in group.}
+#'   \item{`n`}{Number of observations in each group.}
 #' }
+#'
+#' With contouring on (`contour = TRUE`), either [stat_contour()] or
+#' [stat_contour_filled()] (for contour lines or contour bands,
+#' respectively) is run after the density estimate is calculated,
+#' and the computed variables are determined by these stats.
 stat_density_2d <- function(mapping = NULL, data = NULL,
                             geom = "density_2d", position = "identity",
                             ...,
                             contour = TRUE,
+                            contour_type = "lines",
+                            contour_var = "density",
                             n = 100,
                             h = NULL,
                             adjust = c(1, 1),
                             na.rm = FALSE,
                             show.legend = NA,
                             inherit.aes = TRUE) {
+  if (isTRUE(contour_type == "bands")) {
+    stat <- StatDensity2dFilled
+  } else {
+    stat <- StatDensity2d
+  }
+
```

Comment:
I'm not particularly fond of this setup. Basically I don't think `contour_type` should be a user facing parameter but simply a quality of `StatDensity2d` and `StatDensity2dFilled`. One of the reasons is that if this parameter is not set explicitly e.g. by doing something weird like `geom_polygon(stat = 'density2dfilled')` the code will behave like `density2d`, but also because having a switch in the constructor that turns a layer into a new type of layer is not really in line with how we usually do it

Reply:
Agreed. I think this is a hold-over from my first attempt at writing this, which muddled everything together and would have required a lot of code duplication for these two stats. Now, with the new setup, I can do it very cleanly and easily. I'll rewrite.

Reply:
Done. There was really no need for `contour_type`.

Note that the failing CI build for R devel has nothing to do with this PR but is due to #3885.